zenoh/src/net/routing/interceptor/authorization.rs

```diff
@@ -0,0 +1,250 @@
+    put: PermissionPolicy,
+    declare_subscriber: PermissionPolicy,
+    declare_queryable: PermissionPolicy,
+}
+
+impl ActionPolicy {
+    fn action(&self, action: Action) -> &PermissionPolicy {
+        match action {
+            Action::DeclareQueryable => &self.declare_queryable,
+            Action::Get => &self.get,
+            Action::Put => &self.put,
+            Action::DeclareSubscriber => &self.declare_subscriber,
+        }
+    }
+    fn action_mut(&mut self, action: Action) -> &mut PermissionPolicy {
+        match action {
+            Action::DeclareQueryable => &mut self.declare_queryable,
+            Action::Get => &mut self.get,
+            Action::Put => &mut self.put,
+            Action::DeclareSubscriber => &mut self.declare_subscriber,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct FlowPolicy {
+    ingress: ActionPolicy,
+    egress: ActionPolicy,
+}
+
+impl FlowPolicy {
+    fn flow(&self, flow: Flow) -> &ActionPolicy {
+        match flow {
+            Flow::Ingress => &self.ingress,
+            Flow::Egress => &self.egress,
+        }
+    }
+    fn flow_mut(&mut self, flow: Flow) -> &mut ActionPolicy {
+        match flow {
+            Flow::Ingress => &mut self.ingress,
+            Flow::Egress => &mut self.egress,
+        }
+    }
+}
+
+pub struct PolicyEnforcer {
+    pub(crate) acl_enabled: bool,
+    pub(crate) default_permission: Permission,
+    pub(crate) subject_map: SubjectMap,
+    pub(crate) policy_map: PolicyMap,
+}
+
+#[derive(Debug, Clone)]
+pub struct PolicyInformation {
+    subject_map: SubjectMap,
+    policy_rules: Vec<PolicyRule>,
+}
+
+impl PolicyEnforcer {
+    pub fn new() -> PolicyEnforcer {
+        PolicyEnforcer {
+            acl_enabled: true,
+            default_permission: Permission::Deny,
+            subject_map: SubjectMap::default(),
+            policy_map: PolicyMap::default(),
+        }
+    }
+
+    /*
+       initializes the policy_enforcer
+    */
+    pub fn init(&mut self, acl_config: &AclConfig) -> ZResult<()> {
+        self.acl_enabled = acl_config.enabled;
+        self.default_permission = acl_config.default_permission;
+        if self.acl_enabled {
+            if let Some(rules) = &acl_config.rules {
+                if rules.is_empty() {
+                    log::warn!("[ACCESS LOG]: ACL ruleset in config file is empty!!!");
+                    self.policy_map = PolicyMap::default();
+                    self.subject_map = SubjectMap::default();
+                } else {
+                    let policy_information = self.policy_information_point(rules)?;
+                    let subject_map = policy_information.subject_map;
+                    let mut main_policy: PolicyMap = PolicyMap::default();
+
+                    for rule in policy_information.policy_rules {
+                        if let Some(index) = subject_map.get(&rule.subject) {
+                            if let Some(single_policy) = main_policy.get_mut(index) {
```

Comment:
To be honest I don’t quite understand why the case of a separate policy is processed separately, could you please explain a bit.

Reply:
As we discussed, this is for handling cases where first value doesn't exist. Modified code to keep it less verbose.